compiler_opt/rl/env.py

```diff
@@ -0,0 +1,373 @@
+
+  def __init__(self, proc, get_scores_fn, module_name, compiled_module_path):
+    self._proc = proc
+    self._get_scores_fn = get_scores_fn
+    self._module_name = module_name
+    self._compiled_module_path = compiled_module_path
+
+  @property
+  def compiled_module_path(self):
+    return self._compiled_module_path
+
+  def get_scores(self, timeout: Optional[int] = None):
+    self._proc.wait(timeout=timeout)
+    return self._get_scores_fn()
+
+
+class InteractiveClang(ClangProcess):
+  """Wrapper around clang's interactive mode."""
+
+  def __init__(
+      self,
+      proc,
+      get_scores_fn,
+      module_name: str,
+      compiled_module_path: str,
+      reader_pipe: io.BufferedReader,
+      writer_pipe: io.BufferedWriter,
+  ):
+    super().__init__(proc, get_scores_fn, module_name, compiled_module_path)
+    self._reader_pipe = reader_pipe
+    self._writer_pipe = writer_pipe
+    self._obs_gen = log_reader.read_log_from_file(self._reader_pipe)
+
+    self._is_first_obs = True
+
+    self._terminal_obs = _TERMINAL_OBS
+    self._terminal_obs[MODULE_NAME_KEY] = module_name
+
+  def _running(self) -> bool:
+    return self._proc.poll() is None
+
+  def get_observation(self) -> OBS_T:
+    if not self._running():
+      return self._terminal_obs
+
+    def _get_step_type():
+      step_type = FIRST_STEP_STR if self._is_first_obs else MID_STEP_STR
+      self._is_first_obs = False
+      return step_type
+
+    try:
+      obs: log_reader.ObservationRecord = next(self._obs_gen)
+
+      tv_dict = {}
+      for fv in obs.feature_values:
+        array = fv.to_numpy()
+        tv_dict[fv.spec.name] = np.reshape(array, newshape=fv.spec.shape)
+      return {
+          OBS_KEY: tv_dict,
+          REWARD_KEY: obs.score if obs.score else 0.0,
+          SCORE_POLICY_KEY: 0.0,
+          SCORE_DEFAULT_KEY: 0.0,
+          CONTEXT_KEY: obs.context,
+          MODULE_NAME_KEY: self._module_name,
+          OBS_ID_KEY: obs.observation_id,
+          STEP_TYPE_KEY: _get_step_type(),
+      }
+    except StopIteration:
+      return self._terminal_obs
+
+  def send_action(self, action: np.ndarray) -> None:
+    assert self._running()
+    data = action.tobytes()
+    bytes_sent = self._writer_pipe.write(data)
+    # Here we use the fact that for common types, the np.dtype and ctype should
+    # behave the same
+    assert bytes_sent == action.dtype.itemsize * math.prod(action.shape)
+    try:
+      self._writer_pipe.flush()
+    except BrokenPipeError:
+      # The pipe can break after we send the last action
+      pass
+
+
+_EPS = 1e-4
+
+
+def compute_relative_rewards(score_a: dict[str, float],
+                             score_b: dict[str, float]) -> dict[str, float]:
+
+  def _reward_fn(a: float, b: float) -> float:
+    return 1.0 - (a + _EPS) / (b + _EPS)
+
+  assert score_a.keys() == score_b.keys()
+  return {key: _reward_fn(score_a[key], score_b[key]) for key in score_a}
+
+
+@contextlib.contextmanager
+def clang_session(
```

Comment:
make this gin configurable, and the MLGOTasks, too. Then the task_factory can just be a `type`. 

You could make clang_path be part of the ctor of the MLGOTask (basically following the CompilationRunner way for this)

I'd also leave command line construction to the MLGOTask implementation, so maybe just ask the task object: here's the tempdir, here's your baseline cmdline (i.e. the one coming from LoadedModuleSpec), and here's the interactive pipe name (None if you don't want that) or the policy path (None if you want interactive) - maybe best to wrap these 2 into a little class enforcing that _both_ is bad.